### ISSUE TITLE
Non JSON serializable objects are not supported

### ISSUE THREAD

**Issue by ramosmike** [SAP/cf-python-logging-support#22]
Hello,

We are running an application which uses Pika library as AMPQ client, the library has a default behaviour of logging objects such channels to the logger.
It was noticed that when using cf_logging, non JSON serializable objects are not supported:

2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR --- Logging error ---
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/logging/__init__.py", line 992, in emit
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     msg = self.format(record)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/logging/__init__.py", line 838, in format
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     return fmt.format(record)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/site-packages/sap/cf_logging/formatters/json_formatter.py", line 23, in format
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     return _encode(record.__dict__)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/site-packages/sap/cf_logging/formatters/json_formatter.py", line 8, in _encode
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     return json.dumps(obj)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/json/__init__.py", line 231, in dumps
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     return _default_encoder.encode(obj)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/json/encoder.py", line 199, in encode
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     chunks = self.iterencode(o, _one_shot=True)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/json/encoder.py", line 257, in iterencode
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     return _iterencode(o, 0)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/json/encoder.py", line 180, in default
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     o.__class__.__name__)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR TypeError: Object of type 'Channel' is not JSON serializable
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR Call stack:
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/threading.py", line 884, in _bootstrap
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     self._bootstrap_inner()
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/threading.py", line 916, in _bootstrap_inner
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     self.run()
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/threading.py", line 864, in run
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     self._target(*self._args, **self._kwargs)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "start_crawler.py", line 19, in crawling_routine
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     channel.publish_texts(downloader.to_text_results(results.country, results.source_name, results.crawled_items), logger)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/app/communication/channel.py", line 14, in publish_texts
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     _publish('text_handler', t, logger)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/app/communication/channel.py", line 43, in _publish
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     channel.close()
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/site-packages/pika/adapters/blocking_connection.py", line 1435, in close
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     self._impl.close(reply_code=reply_code, reply_text=reply_text)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR   File "/home/vcap/deps/0/python/lib/python3.6/site-packages/pika/channel.py", line 529, in close
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR     reply_code, reply_text, self)
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR Message: 'Closing channel (%s): %r on %s'
   2018-05-23T10:12:27.24-0300 [APP/PROC/WEB/0] ERR Arguments: (0, 'Normal shutdown', <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('10.239.0.74', 47372)->('10.11.241.26', 37968) params=<URLParameters host=10.11.241.26 port=37968 virtual_host=/ ssl=False>>>)

Does cf_logging assumes that all objects to be logged should be JSON serializable?

Thanks,
Micael.

**Comment by adib** [comment]
@desgovedarska – comments?

**Comment by gvachkov** [comment]
Hi ramosmike,

I believe your issue is related to https://github.com/SAP/cf-python-logging-support/issues/12. We have provided fix and I have recently released a new version of the library v3.2.0. 

Have in mind that with the latest version there are changes in logging, where few log fields related to user information are not populated by default. Check the [readme](https://github.com/SAP/cf-python-logging-support#logging-sensitive-data) for details.
